### ISSUE TITLE
Lightweight NLP classifier for secret detection — opt-in, offline, deterministic

### ISSUE THREAD

**Issue by myselfsiddharth** [myselfsiddharth/Flecto#136]
`src/secrets.js` buys its **zero-false-positive floor with recall**, and the header says so plainly: on the measured corpus, 0 false positives, but **~5% of random tokens missed** (~7% at length 24), and **standard-base64 secrets containing `/` are missed by construction** — the charset gate excludes `/` so that hostnames, URLs, and paths can never be candidates.

That trade is correct as a default. Masking a real hostname in someone's terminal is worse than missing an unusual secret. But it leaves a known, quantified gap, and the gap sits exactly where a small text classifier beats fixed thresholds: deciding whether a short opaque string is key material or a word-shaped identifier.

This proposes that classifier as an **opt-in** second-stage detector — the "lightweight NLP model" — with the boundaries drawn tightly enough that it cannot become a liability.

## Proposal

```bash
flecto ci "config/**/*.yaml" --mask-secrets --nlp
flecto watch config/prod.yaml --mask-secrets --nlp --nlp-threshold 0.9
```

`.flectorc`:

```json
{ "defaults": { "nlp": true } }
```

Kill switch for a runner that must not run it at all: `FLECTO_NLP=0`, overriding both.

Off by default in 3.x. On by default is a 4.0 conversation, and only if the benchmark below says it earned it.

## It may only add detections, never remove them

The classifier runs **after** the existing detectors, and its output is **unioned** with theirs:

- Known vendor token format matches → secret. The classifier is not consulted and cannot override it.
- Entropy gate passes → secret. The classifier cannot un-flag it.
- Entropy gate rejects → the classifier gets a vote, and may flag it.

A model that can suppress a detection is a model that can leak a credential on a version bump. Union-only also bounds the failure mode: a bad model costs false positives, which are visible and annoying, rather than silent exposure. These matches should carry their own `SecretKind` (`'classified'`) so the renderer, `SecretMatch`, and the redaction path can distinguish them — and so anyone who doesn't trust it can see precisely what it caught.

## Constraints

**Offline. No network, ever.** Flecto reads `.env` files, SOPS documents, and Terraform plans carrying sensitive values; #113 was a sensitive-value leak. Any design that sends config content to a hosted model is out, including "just the hashes". Local inference over shipped weights, no telemetry, no phone-home.

**Deterministic and pinned.** Teams gate merges on Flecto's exit code. Identical input must produce an identical finding on every run and every machine — no sampling, nothing that lets a float wobble across the threshold. Weights are versioned, the version belongs in the JSON envelope next to `schema_version`, and **changing the model is a breaking change**, treated with the same semver care as changing a policy pack default.

**Small, and no new heavy dependency.** The README promises that nothing extra has to exist on the CI runner, and the tree is 7 small deps. No `onnxruntime`, no `transformers.js`, no wasm blob. What actually fits: **character n-gram logistic regression or naive Bayes, trained offline, exported as a plain JSON weight file, with inference as a dot product in `src/`**. Sub-millisecond, zero new deps, auditable in a diff — and genuinely the right model class for classifying short opaque strings. Budget: **< 500 KB** added to the published package, **< 1 ms** per value.

**`--nlp` toggles the shipped model; it does not load one from disk.** If a `--nlp-model <path>` ever exists, it is CLI-only and never honored from `.flectorc`, for the reason spelled out in GHSA-wq8m-fc3q-8m5x: `.flectorc` is attacker-controlled on an untrusted PR, and deserializing an attacker-supplied model file is the same class of hole as loading their plugin. The simplest safe answer for v1 is to ship exactly one model and offer no path option at all.

## The hard part: training data

There is no corpus to train on, and **we cannot collect one** — user configs are the last thing this project should be gathering. Everything has to come from synthesis and public sources:

- **Positives** — generated key material across the encodings that actually appear (base64, base64url, base62, hex, base32), deliberately including the `/`-containing standard-base64 shapes the charset gate cannot see.
- **Negatives** — the residual false-positive classes the current header already names (`CustomerSuccessDashboard2026`, `AcmeVpnGateway01Prod`, `PRODUCTION_us_east_1_Config2024`), plus UUIDs, SRI digests, git SHAs, image refs, ARNs, ISO-8601 stamps, and real config values from permissively licensed public repos.
- The existing `bench/fixtures.js` and test corpus.

Provenance and license for every source gets documented. The generator and training script live in the repo (`training/`, unshipped), and retraining must be reproducible from a clean checkout. For a security tool, a weight file nobody can regenerate is not acceptable.

## Ship criteria

Extend `bench/` with a labeled precision/recall benchmark, and hold the model to the bar the heuristic already set:

- **False positives on the benign corpus: still 0.** This is the entire reason the gates are conservative. A classifier that trades the FP floor for recall is a regression, not a feature.
- **Recall measurably better** than the heuristic's — the ~5% miss rate and the `/`-base64 blind spot are the targets.
- Numbers published in `docs/`, per class, with the corpus described. "It's ML" is not a claim; a confusion matrix is.

If it cannot clear the first bullet, close this issue and keep the heuristic. That is a genuinely possible outcome: the current gates are well-tuned, and beating 0 FP is a high bar.

## Possible second consumer

Policy packs match key names with hardcoded regexes (`(secret|token|password|api[_-]?key|…)`). The same classifier shape could infer a key's *role* — capacity knob, security toggle, diagnostic flag, endpoint — so that a pack generalizes to `maxPoolSz` or `db.conn.upper_bound` without someone adding a pattern. Out of scope here; worth its own issue if the secret-detection model lands and the machinery proves itself.

## Out of scope

- Any hosted model or LLM call, under any flag.
- Natural-language *generation*. The "plain English" output is templated (`actionNote()` in `src/terraform.js`, `src/renderer.js`) and should stay that way: deterministic, reviewable, translatable, and incapable of hallucinating a wrong pool size into a report.
- Shipping this on by default.

## Naming

`--nlp` is what this gets called in conversation, but the thing proposed is a character n-gram classifier over opaque strings — it does not process natural language. `--classify-secrets` or `--smart-secrets` describes it accurately and won't read as an overclaim to anyone who opens `src/`. Worth settling before it reaches a released flag name, since renaming one afterward is a breaking change.

**Comment by myselfsiddharth** [comment]
## Built it, measured it, and it does not clear the bar

Closing with the negative result recorded, because this issue explicitly anticipated it:

> If it cannot clear the first bullet, close this issue and keep the heuristic. That is a genuinely possible outcome: the current gates are well-tuned, and beating 0 FP is a high bar.

The work is in the closed PR #189 if anyone wants the corpus, the trainer, or the benchmark.

## What was built

Exactly what this issue specified: character 3-gram logistic regression plus 22 named shape features, trained offline from a seeded synthetic corpus, shipped as a 48 KB JSON weight file, inference as a dot product. No new dependency, ~0.013 ms per value, union-only, deterministic, offline, no model-loading flag. All the constraints held.

## The result

| | Secrets caught | Benign flagged |
|---|---|---|
| Heuristic alone | 240/500 | 2/36 |
| With classifier | 241/500 | 2/36 |
| **Classifier's contribution** | **+1** | +0 |

**+1 detection in 500.** Not worth a feature.

## Why — and this is the part worth keeping

An earlier iteration measured **+78 detections and 0 false positives**, taking `/`-bearing base64 from 0/189 to 76/189. That number was wrong in two ways, both found in review:

**1. The benign corpus was missing the hard cases.** It had UUIDs, git SHAs, image refs, ARNs, hostnames — but not the opaque *public identifiers* real configs are full of. An adversarial corpus of ~300 realistic values found the classifier flagging:

- `req_01HQ9Z8X7YB4MNPKRTVWC3DGFJ` — the `<prefix>_<ULID>` format used by Stripe, Segment, Temporal and most modern REST APIs. **12.5% false-positive rate** over 400 such ids.
- `3/L4kqtJlcpXroDVBH40Nrjfkd` — an S3 versionId, verbatim from the AWS docs.
- IPFS CIDs, Firebase push keys, Google Place IDs.

**2. The threshold was circular.** It was derived from the same benign set it was then reported against, so "0 false positives" was true by construction rather than measured. Fitting on a separate calibration set and reporting on an untouched holdout is what exposes the real number.

With both fixed, the honest picture is that the model works fine — **AUC 0.944** — but the distributions overlap exactly where it matters:

```
IPFS CID       Qm…      0.954
prefixed id    pi_…     0.926
Place ID       ChIJ…    0.919
                        ------  threshold must sit above these
secret median            0.809
```

**Opaque high-entropy public identifiers are not distinguishable from opaque high-entropy secrets by their characters alone.** This is the same judgement this issue's own negatives list already made for hex-vs-git-SHA and base32-vs-TOTP-seed — it is just less obvious for base62 and base64, because "it looks random" feels like signal when it is the property *both* classes share.

The `/`-bearing base64 gap is the sharpest case: an S3 versionId is drawn from the same distribution as the secrets the feature existed to catch. Training on one as a negative and the other as a positive is labelling one distribution both ways; when tried, recall went to zero, as it should.

## What would actually be needed

The value alone is not enough information. What separates `db.conn: <base64>` from `version_id: <base64>` is **not in the value** — it is in the key name, the file, and the surrounding config. A model with that context might work. That is a different feature from the one specified here, and it should be its own issue with its own evidence, not a revival of this one.

## Incidental finding

The existing heuristic already false-positives on two values in that corpus — `ChIJN1t_tDeuEmsRUsoyG83frY4` (Google Place ID) and an IPFS CID both clear the entropy gate. Filed separately; the "0 false positives" in `src/secrets.js`'s header is corpus-dependent in the same way this exercise was.

**Comment by myselfsiddharth** [comment]
Closing per the ship criterion in the issue body: the classifier does not beat the heuristic at zero false positives once the benign corpus is realistic. Full result in the comment above; the corpus, trainer and benchmark are in #189 if they are ever wanted.
